### PR TITLE
Expand the SyntaxTokenCache to contain more than just whitespace tokens.

### PR DESCRIPTION
**Pull request by ToddGrun**
Costs below are in allocated bytes typing 26 chars in a large file (percentages are percentage of allocations under ParseDocument for that particular method)

without changes: 
  HtmlMarkupParser.ParseDocument: 280,195,072
    CSharpTokenizer.CreateToken:   39,543,650 (14.1%)
    HtmlTokenizer.CreateToken:     16,135,780 ( 5.8%)

with changes: 
  HtmlMarkupParser.ParseDocument: 247,446,272
    CSharpTokenizer.CreateToken:   21,800,900 (8.8%)
    HtmlTokenizer.CreateToken:     12,079,490 (4.9%)

### CODE REVIEW

**Review by NTaylorMullen — APPROVED** [review]
Looks good to me but I'd wait for sign off from @ajaybhargavb since he's our resident parser expert 😄

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/InternalSyntax/SyntaxTokenCache.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+            return false;
+        }
+
+        public static SyntaxToken GetCachedToken(SyntaxKind kind, string content)
+        {
+            var hash = (kind, content).GetHashCode();
+
+            // Allow the upper 16 bits to contribute to the index
+            var indexableHash = hash ^ (hash >> 16);
```
Comment (ajaybhargavb) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/InternalSyntax/SyntaxTokenCache.cs:58]: Why this?

Comment (ajaybhargavb) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/InternalSyntax/SyntaxTokenCache.cs:58]: And is this unrelated to `CacheSizeBits` which is also 16?

**Review by ajaybhargavb — APPROVED** [review]
Just one question.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/InternalSyntax/SyntaxTokenCache.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/InternalSyntax/SyntaxTokenCache.cs:58]: The hashing calculation gives a 32 bit result. This allows me to let all 32 bits contribute to the calculated index. It's not really related to CacheSizeBits (other than the fact that CacheSizeBits is less that 32 and indicates that we need to do this to use the additional hashing data)

---
In reply to: [437672778](https://github.com/dotnet/aspnetcore/pull/22686#discussion_r437672778) [](ancestors = 437672778)

**Pull request merged**